stablehlo/transforms/LegalizeVhloToVersion.cpp

```diff
@@ -0,0 +1,205 @@
+
+#include <climits>
+#include <memory>
+#include <utility>
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "stablehlo/dialect/Version.h"
+#include "stablehlo/dialect/VhloOps.h"
+#include "stablehlo/transforms/Passes.h"
+#include "stablehlo/transforms/TypeConversion.h"
+
+#define DEBUG_TYPE "compat-passes"
+
+namespace mlir {
+namespace vhlo {
+#define GEN_PASS_DEF_VHLOTOVERSIONPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+///////////////////////
+/// VHLO To Version ///
+///////////////////////
+
+namespace {
+
+struct VhloToVersionPass
+    : public impl::VhloToVersionPassBase<VhloToVersionPass> {
+  VhloToVersionPass() : impl::VhloToVersionPassBase<VhloToVersionPass>() {}
+  VhloToVersionPass(VhloToVersionPassOptions const& opts)
+      : impl::VhloToVersionPassBase<VhloToVersionPass>(opts) {}
+
+  FailureOr<Version> parseTargetVersion(llvm::StringRef versionRef) {
+    if (versionRef == "current") {
+      return VhloDialect::getCurrentDialectVersion();
+    }
+    return Version::get(versionRef);
+  }
+
+  FailureOr<Version> validateTargetVersion(llvm::StringRef versionRef) {
+    auto failOrVersion = parseTargetVersion(versionRef);
+    if (failed(failOrVersion)) {
+      if (targetVersion.empty()) {
+        return emitError(getOperation()->getLoc())
+               << "No target version specified. Specify target using: "
+                  "--vhlo-to-version='target=[targetVersion]'\n"
+               << "Target version must be of the form #.#.# or 'current'.";
+      }
+      return emitError(getOperation()->getLoc())
+             << "Invalid target version number argument '" << targetVersion
+             << "'\n"
+             << "Target version must be of the form #.#.# or 'current'.";
+    }
+
+    Version targetVersion = *failOrVersion;
+    if (targetVersion < VhloDialect::getMinimumDialectVersion()) {
+      return emitError(getOperation()->getLoc())
+             << "target version " << targetVersion.getAsArray()
+             << " is less than minimum supported";
+    }
+    if (VhloDialect::getCurrentDialectVersion() < targetVersion) {
+      return emitError(getOperation()->getLoc())
+             << "target version " << targetVersion.getAsArray()
+             << " is greater than current version";
+    }
+    return targetVersion;
+  }
+
+  void runOnOperation() override {
+    ConversionTarget target(getContext());
+
+    // Validate version number
+    auto failOrVersion = validateTargetVersion(targetVersion);
+    if (failed(failOrVersion)) {
+      return signalPassFailure();
+    }
+    Version targetVersion = *failOrVersion;
+
+    // An op is legal if the target version is in the ops `[min, max]`
+    // supported version range.
+    // Example:
+    //   CustomCallv1 0.0.0 -> 0.0.x
```

Comment:
Capitalize "v" here and below? (Similarly to how op names are spelled right now).